docs/proposals/20220125-ipam-integration.md

```diff
@@ -0,0 +1,410 @@
+   Pool LocalObjectReference `json:"pool,omitempty"`
+ 
+   // Address is the IP address.
+   Address string `json:"address"`
+ 
+   // Prefix is the prefix of the address.
+   Prefix int `json:"prefix,omitempty"`
+ 
+   // Gateway is the network gateway of network the address is from.
+   Gateway string `json:"gateway,omitempty"`
+}
+ 
+// IPAddress is a representation of an IP Address that was allocated from an IP Pool.
+type IPAddress struct {
+   metav1.TypeMeta   `json:",inline"`
+   metav1.ObjectMeta `json:"metadata,omitempty"`
+ 
+   Spec IPAddressSpec `json:"spec,omitempty"`
+}
+```
+
+### Pools & IPAM Providers
+
+In order for IPAM providers to fulfil IPAddressClaims, they likely require a few parameters, e.g. the network from which to allocate an IP address. If there are multiple different IPAM providers, they also need to know which claims they are supposed to fulfil, and which to ignore.
+
+Therefore, IPAM providers are required to bring a custom IP Pool resource. The resource can have an arbitrary structure with the specific parameters for the IPAM system in use, and by inspecting the API group and kind of the pool reference held by IPAddressClaims, the provider can determine whether it is supposed to handle the claim or not.
+
+When an IPAM provider finds a IPAddressClaim that references one of its Pools, it allocates an IP address, creates an IPAddress resource and updates the status of the IPAddressClaim with a reference to this IPAddress.
+
+A simple example for an IP Pool could look like this:
+
+```yaml
+apiVersion: ipam.cluster.x-k8s.io/v1alpha1
+kind: InClusterIPPool
+metadata:
+  name: some-pool
+spec:
+  pools:
+    - subnet: 10.10.10.0/24
+      start: 10.10.10.100
+      end: 10.10.10.200
+```
+A pool for an Infoblox Provider could look like this:
+
+```yaml
+apiVersion: ipam.cluster.x-k8s.io/v1alpha1
+kind: InfobloxIPPool
+metadata:
+  name: ib-pool
+spec:
+  networkView: "some-view"
+  dnsZone: "test.example.com."
+  network: 10.10.10.0/24
+```
+
+### Consumption
+
+The consumers of this API contract are the infrastructure providers. As mentioned in the introduction, not all providers will require integration with IPAM systems. In addition, DHCP may be a viable or even better option in some environments. Integration with this API should therefore be implemented as an optional feature.
+
+In order to consume IPAddresses, infrastructure providers need to create an IPAddressClaim for each address they require, and then have to wait until all claims are fulfilled. When creating those claims, a deterministic naming scheme should be used, so it's easy to identify which claims belong to which machine, e.g. `<infra machine name>-<interface name>`.
+
+Since IPAddressClaims have to reference a Pool, the infrastructure provider needs to allow specifying a Pool in its infrastructure templates, and the referenced Pool also has to be created (or exist already) when creating a cluster. The pool should be in the same namespace.
```

Comment:
>The pool should be in the same namespace.

I think it must since we are relying on local refs. We should probably include some rbac guidance in the proposal for the new CRs.

Reply:
If it's supposed to be in the same namespace, that should be restricted by the type of reference used, which would then be one that simply doesn't allow to specify a namespace.